torchtune/modules/tokenizers/_utils.py

```diff
@@ -4,58 +4,182 @@
+        [1, 1788, 2643, 13, 1792, 9508, 13, 465, 22137, 2933, 2]
+
+
+    Args:
+        messages (List[Message]): A list of messages, each containing role, content,
+            and masked attributes.
+        max_seq_len (Optional[int]): A max sequence length to truncate tokens to.
+            Default: None
+
+    Returns:
+        Tuple[List[int], List[bool]]: The tokenized messages
     """
-    current_slice_len = 0
-    current_slice_is_space = s[0].isspace() if len(s) > 0 else False
-    slice_start = 0
+    start_of_turn = True
+    end_of_turn = False
+    prev_ends_with_space = False
+    tokenized_messages = []
+    mask = []
+    for message in messages:
+        # If assistant message, this is the end of a turn
+        end_of_turn = message.role == "assistant"
 
-    for i in range(len(s)):
-        is_now_space = s[i].isspace()
+        # Prepend BOS on start of new turns
+        if start_of_turn:
+            tokenized_messages.append(bos_id)
+            mask.append(message.masked)
 
-        if current_slice_is_space ^ is_now_space:
-            current_slice_len = 1
-            current_slice_is_space = is_now_space
+        # We want to trim leading whitespace on the next message when
+        # (a) it is a continuation of the turn (i.e. not the first message)
+        # (b) the vocabulary explicitly encodes whitespace characters, and
+        # (c) the previous message did not end with a space
+        trim_leading_whitespace = (
+            (not start_of_turn) and encodes_whitespace and not prev_ends_with_space
+        )
+
+        # Tokenize current message, append with masks
+        tokens = tokenizer.encode(
+            message.content.rstrip(" "),
+            add_bos=False,
+            add_eos=False,
+            trim_leading_whitespace=trim_leading_whitespace,
+        )
+        prev_ends_with_space = message.content.endswith(" ")
+        tokenized_messages.extend(tokens)
+        mask.extend([message.masked] * len(tokens))
+
+        # If assistant message, append EOS at end
+        if end_of_turn:
+            tokenized_messages.append(eos_id)
+            mask.append(message.masked)
+            end_of_turn = False
+            start_of_turn = True
         else:
-            current_slice_len += 1
-            if current_slice_len > max_consecutive_slice_len:
-                yield s[slice_start:i]
-                slice_start = i
-                current_slice_len = 1
-    yield s[slice_start:]
+            start_of_turn = False
+
+        # Break out early if we reach max_seq_len
+        if max_seq_len and len(tokenized_messages) >= max_seq_len:
+            break
+
+    # Finally, truncate if necessary
+    if max_seq_len:
+        tokenized_messages = truncate(tokenized_messages, max_seq_len, eos_id)
+        mask = truncate(mask, max_seq_len, message.masked)
+
+    return tokenized_messages, mask
+
+
+def parse_hf_tokenizer_json(tokenizer_json_path: str) -> Dict[str, int]:
```

Comment:
Is this used anywhere right now? Or just including it for future use? If the latter - while I do like it - maybe leave for when we need it.

Reply:
This is currently used in llama3 and phi3